File: language/src/main/java/language/nodes/CrRootNode.java
package language.nodes;

import com.oracle.truffle.api.frame.*;
import com.oracle.truffle.api.nodes.RootNode;
import language.CrLanguage;
import language.nodes.expr.ExprNode;

public class CrRootNode extends RootNode {
    @Child
    private ExprNode bodyNode;
    private final MaterializedFrame capturedEnv;

    public CrRootNode(CrLanguage language, FrameDescriptor frameDescriptor, ExprNode bodyNode, MaterializedFrame capturedEnv) {
        super(language, frameDescriptor);
        this.bodyNode = bodyNode;
        this.capturedEnv = capturedEnv;
    }

    public CrRootNode(CrLanguage language, FrameDescriptor frameDescriptor, ExprNode bodyNode) {
        this(language, frameDescriptor, bodyNode, null);
    }

    @Override
    public Object execute(VirtualFrame frame) {
        // TODO: extract to Node object
        if (capturedEnv != null) {
            capturedEnv.getFrameDescriptor().getSlots().forEach((slot) -> {
                switch (capturedEnv.getFrameDescriptor().getFrameSlotKind(slot)) {
                    case Long: {
                        long value = FrameUtil.getLongSafe(capturedEnv, slot);
                        frame.getFrameDescriptor().setFrameSlotKind(slot, FrameSlotKind.Long);
                        frame.setLong(slot, value);
                        break;
                    }
                    case Object: {
                        Object value = FrameUtil.getObjectSafe(capturedEnv, slot);
                        frame.getFrameDescriptor().setFrameSlotKind(slot, FrameSlotKind.Object);
                        frame.setObject(slot, value);
                        break;
                    }
                }
            });
        }

        return bodyNode.executeGeneric(frame);
    }
}

File: language/src/main/kotlin/language/nodes/CrRootNode.kt
package language.nodes

import com.oracle.truffle.api.frame.*
import com.oracle.truffle.api.nodes.RootNode
import language.CrLanguage
import language.nodes.expr.ExprNode

class CrRootNode constructor(
    language: CrLanguage,
    frameDescriptor: FrameDescriptor,
    @field:Child private var bodyNode: ExprNode,
    private val capturedEnv: MaterializedFrame? = null
) : RootNode(language, frameDescriptor) {
    override fun execute(frame: VirtualFrame): Any {
        capture(capturedEnv, frame)
        return bodyNode.executeGeneric(frame)
    }

    companion object {
        private fun capture(
            capturedEnv: MaterializedFrame?,
            frame: VirtualFrame
        ) {
            capturedEnv?.frameDescriptor?.slots?.forEach { slot: FrameSlot ->
                when (capturedEnv.frameDescriptor.getFrameSlotKind(slot)!!) {
                    FrameSlotKind.Int -> {
                        val value = FrameUtil.getIntSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Int)
                        frame.setInt(slot, value)
                    }
                    FrameSlotKind.Long -> {
                        val value = FrameUtil.getLongSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Long)
                        frame.setLong(slot, value)
                    }
                    FrameSlotKind.Boolean -> {
                        val value = FrameUtil.getBooleanSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Boolean)
                        frame.setBoolean(slot, value)
                    }
                    FrameSlotKind.Byte -> {
                        val value = FrameUtil.getByteSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Byte)
                        frame.setByte(slot, value)
                    }
                    FrameSlotKind.Double -> {
                        val value = FrameUtil.getDoubleSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Double)
                        frame.setDouble(slot, value)
                    }
                    FrameSlotKind.Float -> {
                        val value = FrameUtil.getFloatSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Float)
                        frame.setFloat(slot, value)
                    }
                    FrameSlotKind.Object -> {
                        val value = FrameUtil.getObjectSafe(capturedEnv, slot)
                        frame.frameDescriptor.setFrameSlotKind(slot, FrameSlotKind.Object)
                        frame.setObject(slot, value)
                    }
                    FrameSlotKind.Illegal -> Unit
                }
            }
        }
    }
}